### DISCUSSION TITLE
CSP without unsafe-inline & docusaurus-theme-classic

### DISCUSSION THREAD

**Discussion by bcaller** [facebook/docusaurus#10231] [Q&A]
docusaurus-theme-classic [injects a script tag](https://github.com/facebook/docusaurus/blob/1f845e681ead1cbee9bbc03c377492a9eeef8f1b/packages/docusaurus-theme-classic/src/index.ts#L130-L144) into the page.

This makes it difficult to apply a Content Security Policy without `script-src 'unsafe-inline'`. Is there a recommended way around this?

Things that could help are switching the injected tag into an external script file, or SHA384 hashing the script content on build and adding the hash to the CSP.

Anyway, just wondered if anyone has solved this before. There doesn't seem to be a nice plugin hook that would work both in dev server and in production builds where the script content can be inspected.

**Comment by slorber** [comment]
This script is inline because we need it to execute asap to avoid flash of unstyled content, notably reading the dark/light mode from localStorage before the UI prints anything.

If we moved this to an external resource, we'd have to block and to wait for it to download before the user sees anything, otherwise the UI would maybe first display the bad theme, and then the theme "gets fixed", producing bad blinking UX. For performance reasons it's much better to inline such scripts, and many other frameworks do the same. It's not just about the theme, but the announcement bar and the classes injected through data attributes too. All this needs to run as soon as possible.

I'm not an expert in CSP. Why do you need that "unsafe-inline" in the first place? Can you share an example of how to have an inline script, and yet make it pass your requirements?

**Reply by bcaller** [reply]
Thanks for the reply.

If you have `Content-Security-Policy` header with `script-src 'self';`, inline scripts will not execute - only those loaded as external scripts from the same origin will execute.

`unsafe-inline` would allow all inline scripts. To provide a layer of defence against XSS attacks I'd prefer to avoid this.

To avoid `unsafe-inline` you can include the hash of the inline script as described at https://content-security-policy.com/hash/

`<script>doSomething();</script>` -> SHA256 of content inside script tags, output as base64 -> `RFWPLDbv2BY+rCkDzsE+0fr8ylGr2R2faWMhq4lfEQc=` -> CSP `script-src 'sha256-RFWPLDbv2BY+rCkDzsE+0fr8ylGr2R2faWMhq4lfEQc=';`

Described at https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src#unsafe_inline_script

> Disallowing ... inline scripts is one of the biggest security wins CSP provides.
> Allowing all inline scripts is considered a security risk, so it's recommended to use a nonce-source or a hash-source instead

It's also possible to use a nonce, but nonces can only be used once i.e. needs to be regenerated on every HTTP request, which means they aren't suitable for statically built sites.

I guess I could make some tool which runs over the built HTML and calculates the hashes on production builds, but not sure how to do that for the dev server build.

**Reply by slorber** [reply]
And how would you exploit XSS on a Docusaurus site in the first place?

That seems quite unlikely to happen considering pages are static and all inserted scripts are injected at build time. The only way probably involves hacking the developer or using a CDN Middleware, in both cases this can be much more harmful?

**Reply by bcaller** [reply]
Yes, it would be due to other elements or scripts added to the page.

**Reply by slorber** [reply]
Sorry but I don't understand your response. 

Can you show an example exploit, as proof that it's worth it for us to invest time in solving this problem? My goal here is to avoid spending time solving theoretical-only problems.

---

> To avoid unsafe-inline you can include the hash of the inline script as described at [content-security-policy.com/hash](https://content-security-policy.com/hash/)

We don't control your server CSP response headers, but you do. Considering the inline script should be the same for all pages, what prevents you from computing the hash manually and hardcoding it as a response of your web server CSP?  

> I guess I could make some tool which runs over the built HTML and calculates the hashes on production builds

Do you really need to make a tool? We don't update that script often and you could just copy/paste the script string manually by running a code snippet on your landing page console?

```js

const scriptContent = document.querySelector("body > script").innerHTML

async function computeScriptCSP(content) {
  const msgUint8 = new TextEncoder().encode(content); 
  const hashBuffer = await window.crypto.subtle.digest("SHA-256", msgUint8);
  const hashArray = Array.from(new Uint8Array(hashBuffer));
  const hashHex = hashArray
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
  return hashHex;
}

computeScriptCSP(scriptContent).then((digestHex) => console.log(`sha256-${digestHex}`));
```

(inspired from https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest#converting_a_digest_to_a_hex_string)

> but not sure how to do that for the dev server build

Why do you need a CSP for the dev server in the first place?

**Reply by sgerace** [reply]
@slorber, I stumbled upon this discussion thread while hardening our internal Docusaurus site. To answer your question:

> Can you show an example exploit, as proof that it's worth it for us to invest time in solving this problem? My goal here is to avoid spending time solving theoretical-only problems.

Though I agree that the exploits are probably going to be mostly theoretical, within the enterprise/corporate world, oftentimes the likelihood of something being exploited does not factor into whether it needs to be addressed. IT departments will often reject a site outright (even if it is only to be used internally) if it does not pass all security criteria. In our case, having a suitably strong CSP configuration is one of those important criteria.

I think including the hash of the inline script in the `script-src` CSP policy (as described in [content-security-policy.com/hash](https://content-security-policy.com/hash/)) is an acceptable solution if removing the inline script altogether is not feasible. If that is the 'official' recommended solution, it would be nice if it was possible to tap into the Docusaurus build process somewhere to obtain the hash value of the inline script so it can be added to the CSP policy. You suggested that: "we don't update that script often and you could just copy/paste the script string manually by running a code snippet on your landing page console." Although that _will_ work for some time (this is what we're doing now), it will inevitably break at some point in the future. At that point, we may not even realize the site isn't fully functioning since many users will not report these sorts of issues. As such, it would be great if we could simply obtain the hash from the build process (I'm thinking an ideal solution would be for us to provide some callback in `docusaurus.config.js`) which we can then incorporate into our downstream servers that control CSP response headers. We could certainly create our own tools to do this (as suggested by @bcaller), but those also might be unreliable since it would have to make a number of assumptions about the expected output.

Quickly answering your last question:

> Why do you need a CSP for the dev server in the first place?

I would agree with you, in my case, this is only an issue for production builds.

**Reply by slorber** [reply]
> it would be nice if it was possible to tap into the Docusaurus build process somewhere to obtain the hash value of the inline script so it can be added to the CSP policy

We have a `postBuild()` plugin lifecycle, and you can run your own code after the build completes. This makes it possible for your code to read the HTML files, compute a hash and add it to some deployment config you have controlling headers.

Our scripts do not change often, so you don't even need to do this automatically. For a given Docusaurus version, just compute the hash manually and hardcode it. When upgrading Docusaurus, the hash will probably not change, and if it does, you can recompute it manually again. You can use error monitoring solutions to know if the hash breaks, or ensure you manually review your site for CSP errors when deploying a new version, or automate this with Playwright and fail the CI on any CSP console error. It's quite visible if your theme doesn't persist dark/light mode anyway, or produce a flash of bad theme on init.

> As such, it would be great if we could simply obtain the hash from the build process

As far as I know there's not a single algo supported for CSP, and you may have one you prefer for your CSP, so exposing a hash for a given algo may not be a solution that works for all persons in your situation. You have the `build/index.html` file, now what prevents you from generating a hash yourself, parsing the HTML file with cheerio in a Node.js script for example?

> We could certainly create our own tools to do this (as suggested by @bcaller), but those also might be unreliable since it would have to make a number of assumptions about the expected output.

Then don't, just copy/paste the code manually. Do you upgrade Docusaurus so often that it's not manageable?

Even if we made the scripts clearly identified with an id for example, you can't guarantee that a hacker will not create a third-party plugin and use the same identifier to conflict with our script. If you want maximum security, you'd rather manually evaluate if the scripts you are whitelisting are not harmful, at least once when the scripts change. You know, supply chain attacks can also affect us, and it's possible that someone publishes a malicious script in my place on npm. How do you protect yourself from "officially harmful Docusaurus inline scripts"?

**Reply by DougReeder** [reply]
I would say the key problem with the workarounds mentioned in this thread, is that they're all **fragile** — they require developers to understand how Docusaurus is doing things differently than a standard static site, and why one can't use a CSP without any "unsafe" directives.  If Docusaurus is a major part of one's work, that's not a problem.  But if the Docusaurus site is ancillary to one's main work, the differences are tribal knowledge, that requires extra effort.

Our team, for one, would like to have a configuration setting for Docusaurus, where we could use a CSP without any "unsafe" directives, even at the expense of some performance.

**Reply by slorber** [reply]
@DougReeder, what makes this fragile exactly?

> they require developers to understand how Docusaurus is doing things differently than a standard static site

We don't! Look at all the other static sites that have a theme toggle feature; they all have an inline script to manage FOUC: Starlight, VitePress, Rspress...

That's a requirement for all sites using a theme, with both SSG and SSR.

> But if the Docusaurus site is ancillary to one's main work, the differences are tribal knowledge, that requires extra effort.

I agree that it would be better if we could make your life easier.

However:
- If you need a CSP in the first place, you probably know what you are doing and how difficult it can be to create a good one
- You probably know that we can not set the `Content-Security-Policy` header for you. At best we can only give you a list of script hashes that you set yourself
- The solution to extract script hashes is not Docusaurus-specific, it's something that would be generic and work on any static deployment using inline scripts


As of today, to host our Docusaurus website with a stricter CSP, we would need that header:

```
"Content-Security-Policy": "script-src 'self' https://www.googletagmanager.com 'sha256-uxBt6cn5c9B9jugocyGCxjRMZxNHC6P+S78Fl6gpmj8=' 'sha256-LF3Z8n3CoT6+StShrU4zdlpOt7gwSfDDuFOnesPW+JI=' 'sha256-a5KkiGzhW4/G+XcbohgQNGbIt/2pk1dJoK6WLvubGtU='"
```

I tried that using our `serve-handler` package (used by `docusaurus serve`) and it seems works fine with the following config:

```json
      headers: [
        {
          "source" : "**/*",
          "headers" : [{
            "key" : "Content-Security-Policy",
            "value" : "script-src 'self' https://www.googletagmanager.com 'sha256-uxBt6cn5c9B9jugocyGCxjRMZxNHC6P+S78Fl6gpmj8=' 'sha256-LF3Z8n3CoT6+StShrU4zdlpOt7gwSfDDuFOnesPW+JI=' 'sha256-a5KkiGzhW4/G+XcbohgQNGbIt/2pk1dJoK6WLvubGtU='"
          }]
        }
      ]
```

At best, we could give you the 3 hashes, printing them somewhere or emitting a `build/csp.json` file that you could have to use to configure your host headers (we can't do this for you).

## Manual solution

I understood this is not what you want to hear, but there's already a convenient way to retrieve those hashes manually: just open the Chrome console with a strict CSP and see the error messages:

<img width="1784" height="1186" alt="CleanShot 2026-05-07 at 10 45 42@2x" src="https://github.com/user-attachments/assets/06369b6f-3579-4067-bd38-2b31e99e10c2" />

If your company cares about strict CSP, they probably have observability tools like Sentry to report when such errors occur in staging/production, too. Or that could be part of your upgrade process.


## Generic solution.

If manual extraction is not an option, the solution to automate this CSP extraction would not be Docusaurus-specific.

If you trust us that the `website/build` output is secure, you would only have to extract the hashes from the static files through a generic solution, one that would work for any static deployment as well.

```bash
npm run docusaurus build
npm run extract-csp-hashes ./build ./build/csp.json
```

For a typical Docusaurus website, all the problematic inline scripts are likely in `build/index.html`, so you could even optimize and only process that specific file. Keep in mind, we don't have control over inline scripts defined by third-party plugins, so we can't strictly guarantee that.

```bash
npm run docusaurus build
npm run extract-csp-hashes ./build/index.html ./build/csp.json
```

Now, are you asking us to build that generic solution for the whole SSG community?

That's something we could consider, but maybe it has already been built in other communities?

## False sense of security?

The goal of CSP is to tighten the security of your site, not simply to blindly comply with a meaningless corporate security policy.

As a static site, you probably want to ensure that your SSG pipeline is not compromised. For example, you want to prevent any first/third-party Docusaurus plugin from injecting a malicious JS script.

Reviewing inline scripts to prevent such issues can be seen as a feature and good security practice, not a bug. If you fully automate the process, then what prevents a third-party plugin from running a Bitcoin miner on your site's visitors after any npm upgrade? 

If you automate the process, you just add a potentially harmful hash to your header without even knowing.

That is why it is important to understand what you want to protect yourself from in the first place with this CSP. A supply chain risk? A compromised host server? Something else?

**Reply by slorber** [reply]
> Our team, for one, would like to have a configuration setting for Docusaurus, where we could use a CSP without any "unsafe" directives, even at the expense of some performance.

If that helps, we could add an option to our classic theme to emit the inline script to a separate file.

However, this would only work on a per-plugin basis, and we can't force third-party plugin authors to do the same and prevent all inline scripts.

For example, our gtag plugin currently uses an inline script too, as shown on the gtag installation guide: https://developers.google.com/tag-platform/gtagjs

**Reply by DougReeder** [reply]
FWIW, light-dark theming can now be done with the [light-dark() CSS function](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Values/color_value/light-dark)

I'm coming from using 11ty, which does not require CSP unsafe- directives or hashes (and, yes, can't do many of the things Docusaurus can do).  We inherited Docusaurus from a previous team — it appears that Docusaurus is a heavier-weight tool than we want to support.

Thanks for the clarification!

**Reply by slorber** [reply]
> FWIW, light-dark theming can now be done with the [light-dark() CSS function](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Values/color_value/light-dark)


This is not widely supported, and doesn't permit implementing a theme toggle button. It only follows the theme of your OS.

To implement a theme toggle button and have this CSS function react to the change, this would require another proposal to land in the spec: https://github.com/WICG/web-preferences-api

---

> I'm coming from using 11ty, which does not require CSP unsafe- directives or hashes (and, yes, can't do many of the things Docusaurus can do). 

Show me an Eleventy website with a theme toggle that doesn't have an inline script tag. If they don't have it, they have a flash of bad theme on load, not much better.

These sites have inline scripts:
- https://www.11ty.dev
- https://eleventy-libdoc.netlify.app/

If you compare Docusaurus with the classic theme VS an empty/homemade Eleventy site, that's not fair. You can also implement your own custom theme in Docusaurus and have no inline script if you really want to.

---

> We inherited Docusaurus from a previous team — it appears that Docusaurus is a heavier-weight tool than we want to support.

Docusaurus builds a React SPA, so it is indeed heavier than MPAs by design. There are pros and cons to each approach.

We are also opinionated toward docs/blog/pages use case, and our theme provides a theme switch and other features requiring inline scripts to work well. 

Those features are not strictly mandatory and we could eventually give an option to disable them on the classic theme.
